graph/patterns/infrastructure-as-code.md

```diff
@@ -0,0 +1,172 @@
+# Infrastructure as code
+
+Microsoft Graph API Design Pattern
+
+*Infrastructure as code (IaC) resource pattern ensures that system resources can be deployed in a reliable, repeatable, and controlled way, normally via idempotent operations.*
+
+## Problem
+
+Infrastructure as code (IaC) defines system resources and topologies in a descriptive manner that allows teams to manage those resources as they would code. 
+Practicing IaC helps teams deploy system resources in a reliable, repeatable, and controlled way. 
+IaC also helps automate deployment and reduces the risk of human error, especially for complex large environments. 
+Customers want to adopt IaC practices for many of the resources managed through Microsoft Graph.
+
+Most resources' creation operations in Microsoft Graph are not idempotent in nature.
+As a consequence, API consumers that want to offer IaC solutions, must create compensation layers that can mimic idempotent behavior. 
+For example, when creating a resource, the compensation layer must check whether the resource first exists, before trying to create or update the resource.
+
+Additionally, IaC code scripts or templates usually employ client-provided names (or keys) to track resources in a predictable manner, whereas [Microsoft Graph guidelines](../GuidelinesGraph.md#behavior-modeling) suggests use of `POST` to create new entities with service-generated keys.
+
+## Solution
+
+The solution is to use an `UPSERT` pattern, to solve for the non-idempotent creation and client-provided naming problems.
+
+* For IaC scenarios, resources must use `UPSERT` semantics with an [alternate key](./alternate-key.md):
```

Comment:
I don't think alternate key calls this out, but we should state that uniqueName or another client provided alternate key is immutable after being set.  If the value is null then it should be updateable as a way to backfill existing objects to use for IaC scenarios.

Reply:
Thanks - forgot to mention that.  That was something I wanted to say.  I need to go back to the other doc I wrote, which I think is more exhaustive.

Reply:
Adding this to the Issues and considerations section.